pip_run/scripts.py

```diff
@@ -151,24 +150,28 @@ def read_python(self):
         >>> DepsReader(r"__requires__='foo\nbar\n#baz'").read()
         ['foo', 'bar']
         """
-        raw_reqs = suppress(ValueError)(self._read)('__requires__') or []
+        raw_reqs = self._read('__requires__', default=())
         reqs_items = jaraco.text.yield_lines(raw_reqs)
         deps = Dependencies.load(reqs_items)
-        with contextlib.suppress(Exception):
-            deps.index_url = self._read('__index_url__')
+        deps.index_url = self._read('__index_url__')
         return deps
 
-    def _read(self, var_name):
+    def _read(self, var_name, default=None):
         mod = ast.parse(self.script)
-        (node,) = (
-            node
-            for node in mod.body
-            if isinstance(node, ast.Assign)
-            and len(node.targets) == 1
-            and isinstance(node.targets[0], ast.Name)
-            and node.targets[0].id == var_name
-        )
-        return ast.literal_eval(node.value)
+        code = None
+        for node in mod.body:
+            if isinstance(node, ast.Assign) and len(node.targets) == 1:
+                (target,) = node.targets
+            elif isinstance(node, ast.AnnAssign):
+                target = node.target
+            else:
+                continue
+            if isinstance(target, ast.Name) and target.id == var_name:
+                if code:
+                    code = None
+                    break
```

Comment:
It took me a while to figure out that this block here is for the duplication detection. It's nested several levels deep and alters other state variables, entangling it with the other concerns in this method. Much better would be to have a `_read_all` (or similar) that generates all of the targets and have that function react on the condition of "more than one", addressing that concern at a higher level rather than tweaking the internal state to simulate a different outcome.

Reply:
I see. But why would it be a different outcome?

Reply:
I'm suggesting the same outcome, just a different approach.